Add istft to jax.scipy.signal.

yotarok opened the pull request:
Following the discussion in #9411, ISTFT is added to signal.py.
(ISTFT wasn't included in the first PR because it doesn't share a common utility function used for STFT.)

Commit:
Add istft to jax.scipy.signal.

jakevdp commented on jax/_src/scipy/signal.py:
```diff
@@ -492,3 +492,110 @@ def welch(x, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
+@_wraps(osp_signal.istft)
+def istft(Zxx, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
+          input_onesided=True, boundary=True, time_axis=-1, freq_axis=-2):
+  # Input validation
```
Call `_check_arraylike` on `Zxx` to start with

yotarok replied:
Done

jakevdp commented on jax/_src/scipy/signal.py:
```diff
@@ -492,3 +492,110 @@ def welch(x, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
+    raise ValueError('Must specify differing time and frequency axes!')
+
+  Zxx = jnp.asarray(Zxx, dtype=jax.dtypes.canonicalize_dtype(
+      np.result_type(Zxx, np.complex64)))
```
Probably not for this PR, but this kind of thing is used frequently enough that we should probably define a `_promote_dtypes_complex` helper similar to our current `_promote_dtypes_inexact`.

yotarok replied:
I agree, such utility function should be useful.

jakevdp commented on jax/_src/scipy/signal.py:
```diff
@@ -492,3 +492,110 @@ def welch(x, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
+
+  # Get window as array
+  if isinstance(window, (str, tuple)):
+    win = osp_signal.get_window(window, nperseg)
```
I think we should call `win = jnp.asarray(win)` to avoid doing the operations below at compile time.

yotarok replied:
Done

jakevdp commented on jax/_src/scipy/signal.py:
```diff
@@ -492,3 +492,110 @@ def welch(x, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
+        time_axis -= 1
+      x = jnp.moveaxis(x, -1, time_axis)
+
+  time = jnp.arange(x.shape[0]) / float(fs)
```
This cast to `float(fs)` will prevent `fs` from being traced by JIT, vmap, etc. - I don't think we need it?

yotarok replied:
No, we don't need it. Deleted.

jakevdp commented on tests/scipy_signal_test.py:
```diff
@@ -376,6 +383,59 @@ def testWelchWithDefaultStepArgsAgainstNumpy(
+    jsp_fun = partial(jsp_signal.istft,
+        fs=fs, window=window, nperseg=nperseg, noverlap=noverlap, nfft=nfft,
+        input_onesided=onesided, boundary=boundary,
+        time_axis=timeaxis, freq_axis=freqaxis)
```
You might consider making `fs` one of the arguments to `osp_fun`/`jsp_fun` so that we test dynamic inputs for that value. If you remove the `float` cast in the implementation, I believe it will work.

yotarok replied:
Done.

jakevdp reviewed (changes requested):
Looks great! A few minor requests. One last comment, to ensure this is part of the documentation we should add it to the list in https://github.com/google/jax/blob/main/docs/jax.scipy.rst

yotarok commented:
Thanks for the feedbacks.
I updated the branch reflecting your comments.
In addition to that, I modified the method for implmenting overlap-add following the suggestion given by Peter (https://github.com/google/jax/issues/9411#issuecomment-1082066504)

yotarok commented:
Fixed.
I first tried to fix this by simulating "osp" more accurately.  But, I found that the behavior is different between the version used in CI (probably 1.2.3?) and the newer version (1.8.0).
So, I gave up and disabled dtype checks.

jakevdp commented:
Sounds good - thanks!
